### ISSUE TITLE
In react fiber, is didReceiveUpdate related to dirty checking?

### ISSUE THREAD

**Issue by Low-ProFiles** [facebook/react#28622]
https://github.com/facebook/react/blob/f09e1599d631051a559974578a6d4c06effd95eb/packages/react-reconciler/src/ReactFiberBeginWork.js#L620

I have some question while looking for ReactFiberBeginWork.js file.
**Is didReceiveUpdate related to dirty checking?**

when ```didReceiveUpdate = true``` called, it has to satisfied below conditions.
```(oldProps !== newProps || hasContextChanged() || (workInProgress.type !== current.type))```

To organize the conditions as I understand them,
1. oldProps !== newProps -> props changed
2. hasContextChanged() -> context changed
3. workInProgress.type !== current.type -> code changed

But I understand that dirty checking only works when there is a change in props. 
But didReceiveUpdate is also called for two conditions other than a change in props.
So I'm wondering if didReceiveUpdate is related to dirty checking, and if not, what's the difference

If you're not busy, it would be a great help and honor to organize and understand the concept. Thanks for your read!

**Comment by AnuragKacham** [comment]
The 1st condition directly relates to the usage of dirty checking, but the other two do not. But, they work in the same way after detecting a change. (Change of context in the component or change of the component in total). So, it is not totally related to dirty checking but it uses dirty checking as a small part.Hope it helped!

**Comment by rickhanlonii** [comment]
Dirty checking is used to determine which individual pieces of state have changed, and then process those updates (so the diff happens first, then the update). React doesn't do dirty checking because it re-renders everything, and then compares the output with the previous output (so the update happens first, then the diff).

While this looks like dirty checking, because it's diffing the props/context, it's not really the same thing. This check is just used to bailout of the "re-render everything" process once we reach child in the tree that couldn't have changed, as a perf optimization, not the mechanism to track updates. 

In other words, the flag is not used to determine _what_ to update, but what _didn't_ update.
